ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c

```diff
@@ -836,6 +916,128 @@ NTSTATUS CreateReparsePoint(POBJECT_ATTRIBUTES poa, LPCWSTR SubstituteName, LPCW
 	return status;
 }
 
+
+/*
+* go through all mointpoints (IOCTL_MOUNTMGR_QUERY_POINTS) and check if our driveletter is in the list
+* return 1 if yes, otherwise 0
+*/
+NTSTATUS mountmgr_is_driveletter_assigned(PDEVICE_OBJECT mountmgr, wchar_t driveletter, BOOLEAN *ret) {
+	MOUNTMGR_MOUNT_POINT point = { 0 };
+	MOUNTMGR_MOUNT_POINTS points;
+	PMOUNTMGR_MOUNT_POINTS ppoints = NULL;
+	int len;
+	*ret = 0;
+	NTSTATUS Status;
+
+	ppoints = &points;
+	Status = dev_ioctl(mountmgr, IOCTL_MOUNTMGR_QUERY_POINTS, &point, sizeof(MOUNTMGR_MOUNT_POINT), ppoints, sizeof(MOUNTMGR_MOUNT_POINTS), FALSE, NULL);
+
+	if (Status == STATUS_BUFFER_OVERFLOW) {
+		len = points.Size;
+		ppoints = kmem_alloc(len, KM_SLEEP);
+		Status = dev_ioctl(mountmgr, IOCTL_MOUNTMGR_QUERY_POINTS, &point, sizeof(MOUNTMGR_MOUNT_POINT), ppoints, len, FALSE, NULL);
+	}
+	dprintf("IOCTL_MOUNTMGR_QUERY_POINTS return %x - looking for driveletter '%c'\n", Status,
+		driveletter);
+	if (Status == STATUS_SUCCESS) {
+		char mpt_name[PATH_MAX];
+		for (int Index = 0; Index < ppoints->NumberOfMountPoints; Index++) {
+			PMOUNTMGR_MOUNT_POINT ipoint = ppoints->MountPoints + Index;
+			PUCHAR DeviceName = (PUCHAR)ppoints + ipoint->DeviceNameOffset;
+			PUCHAR SymbolicLinkName = (PUCHAR)ppoints + ipoint->SymbolicLinkNameOffset;
+
+			// Why is this hackery needed, we should be able to lookup the drive letter from volume name
+			dprintf("   point %d: '%.*S' '%.*S'\n", Index,
+				ipoint->DeviceNameLength / sizeof(WCHAR), DeviceName,
+				ipoint->SymbolicLinkNameLength / sizeof(WCHAR), SymbolicLinkName);
+
+			ULONG len = 0;
+			RtlUnicodeToUTF8N(mpt_name, MAXPATHLEN, &len, SymbolicLinkName, ipoint->SymbolicLinkNameLength);
+			mpt_name[len] = 0;
+			char c_driveletter;
+			wctomb(&c_driveletter, driveletter);
+			if (MOUNTMGR_IS_DRIVE_LETTER_A(mpt_name) && mpt_name[12] == c_driveletter) {
+				*ret = 1;
```

Comment:
Do we know who/what releases the `kmem_alloc` above in this positive exit branch? Is it one of those "Windows will free it?" if so, we should probably not use kmem. If we are to free it, where do we?

Reply:
oh yeah, that should be freed before return 
will do that